2023 county-level delinquent debt overall and by auto loan, medical debt, and student loan debt created

minglizhong opened the pull request:
Closes #406 

All 2023 results are in mobility-from-poverty\01_financial-well-being\output. Previously created 2021 city-level data and 2022 county-level data are also in the same \output subfolder. The do file that needs to be reviewed is this: mobility-from-poverty\01_financial-well-being\county-debt-coll-shares-2023.do, mostly copied from the previous do file (county-debt-coll-shares-2022.do) with some revisions. 
Note that in the previous 2022 county-level data, mobility-from-poverty\01_financial-well-being\output\county-debt-coll-shares-2022.csv, state and county names are numeric (1, 2, 3...). I don't know the actual values of these numeric values. In the new 2023 county-level data, I temporarily kept the state and county actual names, so they're string variables. Further guidance is needed whether we use numeric or string and how numeric values correspond with string values if we'd like to merge 2022 and 2023 county-level data.

kmartinchek commented:
Great work! Some comments from my review of the draft PR (focusing just on the 2023 file). The file ran without errors, but I have several suggestions:

1. IIRC, only final data will be stored in the repo, and directions for download from Box should be incorporated (see https://github.com/UI-Research/mobility-from-poverty/wiki/6_Raw-Data for details). I think you’ll need to edit this file to reflect this guidance and add details on the data. Alternatively, I also think you will be able to download the DiA data from the Data Catalog directly, but I would reach out to Sam Cressman to confirm.
2. You’ll need to note that if using direct file paths on line 23 and line 72 and line 112, for users to replace these. Or, I recommend setting a macro for the first part of the file path so it is relative – which can be done in a “file setup” section at the top of the do file.
3. I would delete lines 25-31 and 149-160 if not needed (they are from prior do file).
4. Add a comment explaining the reason for the drop on line 40 and 81 and 121.
5. I would confirm with Claudia/JR, but I was expecting that each debt variable would be a separate variable (with distinct names and their own quality flag) in one output file – e.g., share_debt_coll, share_debt_med, share_debt_auto, and share_debt_stud. Just because debt in collections is a specific designation (and the other forms of debt don’t meet this same criteria – they are XX # of days past due vs. reflecting lenders acting on derogatory status). As of now, the output files are set up as one for each debt type (using the same variable name), which doesn’t match the data expectations guidelines on the Wiki.
6. I would look closely at the Wiki on data structure, as the state/county variables are names in the current output vs. FIPS codes – but maybe this is okay with Claudia/JR. There should also be an overall file separate from the subgroup file, which was not generated that should be.
7. The file has a lot of repetitive code. I think it could make sense to make changes to the code so that you are pulling in the files and making the rename adjustments first (which could be made easier if you are using distinct variable names for each form of debt), then pulling the updated files into a loop where you do the other formatting and variable generation. 
8. The data quality flag needs a detailed explanation of the 3 flags in DiA data (the n/a’s) so folks know why these are flagged as data quality 3 – but it is possible that these should be flagged as missing because the main metric is NA (see: https://github.com/UI-Research/mobility-from-poverty/wiki/2_Data-Structure-and-Expectations%E2%80%AF#missing-values) , although I would confirm this specifically with Claudia/JR.
9. Also, should the final output have NA values as blank or as “NA”. I would confirm and adjust accordingly.
10. IIRC, the final output should include all years of data (confirm with Claudia/JR)—so you might need to append the 2018, 2022, and 2023 data in this do file or another.

minglizhong commented:
@kmartinchek, thanks so much for your detailed feedback. I've addressed the feedback from your previous comments. I've also consulted with the leadership team. All problems have been resolved.

cdsolari commented:
@minglizhong Thanks so much for generating the debt and medical debt data and for the race-ethnic subgroup-areas. You did a great job of addressing all the comments. One thing I noticed in the output files was that you have cases where the value of the metric is "NA", but it has a data quality value (e.g., 3). We do have guidance [HERE](https://github.com/UI-Research/mobility-from-poverty/wiki/2_Data-Structure-and-Expectations%E2%80%AF#missing-values) on that in case that is written more clearly. Please let me know if you have any questions. Otherwise, I think this looks good! Thanks!

minglizhong commented:
Per Claudia's feedback, the metric quality values have been changed to NA if the metric value is NA in all four csv files.

kmartinchek commented:
Comments on most recent run:
1. The global raw on line 29 and the path on line 44 are incompatible and result in an error unless you remove "dia-2023" from the path or instruct the user to add them to a specific folder. This should be addressed.
2. Medical debt in collections is available in the overall data file you pull initially-- why not clean it from there? It is fine as-is but in future updates, could help users from needing to download multiple data files.
3. Remove lines 94 through 178 and keep in a separate place for Urban reference, as this code is meant to be public-facing. This should be addressed.
4. I would edit the instructions lines with the following: (a) make clear this only shows 2023 data -- line 6, (b) the instructions in line 10 should direct users where and how to download the raw data or access our Box folder where it is held -- right now it just tells the users your file path used.
5. I mostly checked output for NA in metric and NA in quality variable matches. This looked good but I recommend using more assert statements in the code so manual checking is a backup. Also recommend adding code that enables you to look at summary stats and distributions to make sure the data looks as expected. This is done in the other metric files, so figured worth suggesting here. @cdsolari has looked at output and confirmed it looked as expected, so I'll defer to her if she wants any of the changes I describe here.

cdsolari commented:
@minglizhong please see the above review by @kmartinchek. In her above comments, I think item number 1 violates our rules for finishing a metric. If you could address that, I think we can flag the rest of the items as room for code efficiency in the future. Please let me know if you are unable to do this. It does seem like the output files otherwise look good, which is most critical at this point. Thank you!

minglizhong commented:
Hi @cdsolari , regarding #1, the paths on line 29 and 44 have no error. The raw data is saved in "Box\Data\Metrics_2025_round\debt_in_collections\dia-2023" with dia-2023 in the path. Kassie's error could be due to her without access to the Box folder or without changing the path to her local path. I had a note in the script at the proper places noting that users need to change to their local paths. Hope it helps.

cdsolari reviewed (approved):
the county fips variable successfully removed the state code, and the final evaluation form passed all the tests. Notes were also successfully added into the body of the code to highlight methodological changes noted in the Debt in America website notes. This all looks good.
